canopycover.py

```diff
@@ -0,0 +1,358 @@
+#!/usr/bin/env python3
+"""Calculates canopy coverage for plots in georeferenced images
+"""
+
+import argparse
+import copy
+import json
+import logging
+import os
+from typing import Union
+import numpy as np
+import dateutil.parser
+from agpypeline import entrypoint, algorithm, geoimage
+from agpypeline.environment import Environment
+from osgeo import gdal, ogr
+
+from configuration import ConfigurationCanopycover
+
+
+# Array of trait names that should have array values associated with them
+TRAIT_NAME_ARRAY_VALUE = ['canopy_cover', 'site']
+
+# Mapping of default trait names to fixed values
+TRAIT_NAME_MAP = {
+    'access_level': '2',
+    'species': 'Unknown',
+    'citation_author': '"Zongyang, Li"',
+    'citation_year': '2016',
+    'citation_title': 'Maricopa Field Station Data and Metadata',
+    'method': 'Green Canopy Cover Estimation from Field Scanner RGB images'
+}
+
+
+def get_fields() -> list:
+    """Returns the supported field names as a list
+    """
+    return ['local_datetime', 'canopy_cover', 'access_level', 'species', 'site',
+            'citation_author', 'citation_year', 'citation_title', 'method']
+
+
+def get_default_trait(trait_name: str) -> Union[str, list]:
+    """Returns the default value for the trait name
+    Args:
+       trait_name: the name of the trait to return the default value for
+    Return:
+        If the default value for a trait is configured, that value is returned. Otherwise
+        an empty string is returned.
+    """
+    # pylint: disable=global-statement
+    global TRAIT_NAME_ARRAY_VALUE
+    global TRAIT_NAME_MAP
+
+    if trait_name in TRAIT_NAME_ARRAY_VALUE:
+        return []  # Return an empty list when the name matches
+    if trait_name in TRAIT_NAME_MAP:
+        return TRAIT_NAME_MAP[trait_name]
+
+    return ""
+
+
+def get_traits_table() -> list:
+    """Returns the field names and default trait values
+
+    Returns:
+        A tuple containing the list of field names and a dictionary of default field values
+    """
+    # Compiled traits table
+    fields = get_fields()
+    traits = {}
+    for field_name in fields:
+        traits[field_name] = get_default_trait(field_name)
+
+    return [fields, traits]
+
+
+def generate_traits_list(traits: list) -> list:
+    """Returns an array of trait values
+
+    Args:
+        traits(dict): contains the set of trait values to return
+
+    Return:
+        Returns an array of trait values taken from the traits parameter
+    """
+    # compose the summary traits
+    fields = get_fields()
+    trait_list = []
+    for field_name in fields:
+        if field_name in traits:
+            trait_list.append(traits[field_name])
+        else:
+            trait_list.append(get_default_trait(field_name))
+
+    return trait_list
+
+
+def setup_default_traits(traits: dict, args: argparse.Namespace, full_md: list) -> dict:
```

Comment:
not clear why we are setting up default traits? the only trait in this case is 'canopy_cover' and it won't have a default value. all of the other

Reply:
This is a naming convention from TERRA REF. See https://github.com/terraref/extractors-stereo-rgb/blob/3efa2dba583cd31bc106a30f8c211711a1fd4762/canopycover/terra_canopycover.py#L24